File: strings/Strings.java
import java.util.ArrayList;
import java.util.Scanner;

public class Strings {

    public int[] prefixFunction(String s) {
        int[] p = new int[s.length()];
        p[0] = 0;
        int k;
        for (int i = 1; i < s.length() - 1; i++) {
            k = p[i - 1];
            while (k > 0 && s.charAt(i) != s.charAt(k)) {
                k = p[k - 1];
            }
            if (s.charAt(i)==s.charAt(k)) {
                k++;
            }
            p[i] = k;
        }
        return p;
    }

    public ArrayList<Integer> kmp(String P, String T) {
        int pl = P.length();
        int tl = T.length();
        ArrayList<Integer> answer = new ArrayList<>();
        int[] p = prefixFunction(P+"#"+T);
        for (int i = 0; i < tl - 1;i++) {
            if (p[pl+i+1] ==pl) {
                answer.add(i);
            }
        }
        return answer;
    }

    public static void main(String[] args) {
        Strings strings = new Strings();
        String substring, string;
        Scanner scanner = new Scanner(System.in);
        substring = scanner.nextLine();
        string = scanner.nextLine();
        ArrayList<Integer> answer = strings.kmp(substring,string);
        if (answer.size() == 0) {
            System.out.print(-1);
        } else {
            System.out.print(answer.get(0)-substring.length()+1);
        }
        System.exit(0);
    }
}

File: strings/strings.py
def prefix_function(s):
    p = [0] * len(s)
    p[0] = 0
    for i in range(1, len(s)):
        k = p[i - 1]
        while k > 0 and s[i] != s[k]:
            k = p[k - 1]
        if s[i] == s[k]:
            k += 1
        p[i] = k
    return p


def kmp(P, T):
    pl = len(P)
    tl = len(T)
    answer = []
    p = prefix_function(P + "#" + T)
    for i in range(0, tl):
        if p[pl + i + 1] == pl:
            answer.append(i)
    return answer


# Символы в строке считаются с нуля
# Ответ: индекс первого символа первого вхождения в строку
T = input()
P = input()
occurrences = kmp(P, T)
if occurrences:
    answer = occurrences.pop(0) - len(P) + 1
else:
    answer = -1
print(answer)
